File: client_app/lib/kill_switch_confirm.dart
/// Kill-switch Settings confirm gate (pure logic, no dialog I/O).
///
/// Enabling kill-switch requires an explicit typed confirmation so users do not
/// arm fail-closed egress by accident. Disabling never requires a phrase.
library;

/// Exact token the user must type to enable kill-switch (case-sensitive).
const String kKillSwitchConfirmToken = 'KILLSWITCH';

/// Dialog title / lead (all-caps warning).
const String kKillSwitchConfirmTitle = 'ARE YOU SURE?';

/// Risk explanation shown when enabling kill-switch.
const String kKillSwitchConfirmRiskBody =
    'Turning kill-switch ON can block all non-VPN traffic if residual drops. '
    'That may break captive portals, app/OS updates, local network access, '
    'and other connections until you turn it OFF or reconnect residual. '
    'Default is OFF for a reason. Type KILLSWITCH below to confirm enable.';

/// Confirm button label on the enable dialog.
const String kKillSwitchConfirmActionLabel = 'Enable kill switch';

/// Cancel button label.
const String kKillSwitchConfirmCancelLabel = 'Cancel';

/// Field hint for the typed confirm box.
const String kKillSwitchConfirmFieldHint = 'Type KILLSWITCH to confirm';

/// Marker key for the enable-confirm dialog shell (widget tests).
const String kKillSwitchConfirmDialogKey = 'kill_switch_confirm_dialog';

/// Marker key for the typed confirm TextField.
const String kKillSwitchConfirmFieldKey = 'kill_switch_confirm_field';

/// Result of evaluating whether a kill-switch opt-in change may persist.
class KillSwitchConfirmDecision {
  const KillSwitchConfirmDecision({
    required this.allowPersist,
    required this.nextOptIn,
    this.reason = '',
  });

  /// True when Settings may save [nextOptIn].
  final bool allowPersist;

  /// Value to persist when [allowPersist] is true.
  final bool nextOptIn;

  /// Why the gate rejected or allowed (tests / logs).
  final String reason;
}

/// Pure gate for kill-switch Settings changes.
///
/// - Turning **OFF** (`desiredOn == false`): always allowed, no token needed.
/// - Turning **ON** (`desiredOn == true`): allowed only when [confirmText]
///   equals [kKillSwitchConfirmToken] exactly (after optional trim of outer
///   whitespace only — internal case must match).
/// - Empty / wrong / cancelled ON attempts: [allowPersist] false, next stays off.
KillSwitchConfirmDecision evaluateKillSwitchConfirm({
  required bool desiredOn,
  String? confirmText,
  bool cancelled = false,
}) {
  if (!desiredOn) {
    return const KillSwitchConfirmDecision(
      allowPersist: true,
      nextOptIn: false,
      reason: 'disable_no_confirm',
    );
  }
  if (cancelled) {
    return const KillSwitchConfirmDecision(
      allowPersist: false,
      nextOptIn: false,
      reason: 'enable_cancelled',
    );
  }
  final typed = (confirmText ?? '').trim();
  if (typed == kKillSwitchConfirmToken) {
    return const KillSwitchConfirmDecision(
      allowPersist: true,
      nextOptIn: true,
      reason: 'enable_token_ok',
    );
  }
  return KillSwitchConfirmDecision(
    allowPersist: false,
    nextOptIn: false,
    reason: typed.isEmpty ? 'enable_empty_token' : 'enable_wrong_token',
  );
}

/// True when [text] is the exact enable token (trim outer whitespace).
bool killSwitchConfirmTokenMatches(String? text) {
  return (text ?? '').trim() == kKillSwitchConfirmToken;
}

File: client/kill_switch_confirm.py
"""Kill-switch Settings confirm gate (pure logic, no dialog I/O).

Mirrors ``client_app/lib/kill_switch_confirm.dart`` so Windows (and other
native clients) share the same enable semantics:

- Enable requires exact typed token ``KILLSWITCH`` (outer whitespace trimmed).
- Disable never requires a phrase.
- Cancel / wrong / empty enable leaves kill-switch **OFF**.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Optional

# Exact token the user must type to enable kill-switch (case-sensitive).
KILL_SWITCH_CONFIRM_TOKEN = "KILLSWITCH"

# Dialog title / lead (all-caps warning).
KILL_SWITCH_CONFIRM_TITLE = "ARE YOU SURE?"

# Risk explanation shown when enabling kill-switch.
KILL_SWITCH_CONFIRM_RISK_BODY = (
    "Turning kill-switch ON can block all non-VPN traffic if residual drops. "
    "That may break captive portals, app/OS updates, local network access, "
    "and other connections until you turn it OFF or reconnect residual. "
    "Default is OFF for a reason. Type KILLSWITCH below to confirm enable."
)

# Confirm / cancel button labels.
KILL_SWITCH_CONFIRM_ACTION_LABEL = "Enable kill switch"
KILL_SWITCH_CONFIRM_CANCEL_LABEL = "Cancel"
KILL_SWITCH_CONFIRM_FIELD_HINT = "Type KILLSWITCH to confirm"

# Settings surface warning chrome (parity with Flutter leak_posture.dart).
KILL_SWITCH_WARNING_TITLE = "WARNING"
KILL_SWITCH_SETTINGS_LABEL = "KILL SWITCH"
KILL_SWITCH_SETTINGS_BODY = (
    "Optional fail-closed firewall when residual is connected. "
    "Default is OFF. Enabling requires typing KILLSWITCH after a clear warning. "
    "If residual drops while kill-switch is ON, non-VPN internet may be blocked "
    "until you turn it OFF or reconnect."
)
KILL_SWITCH_ENABLE_SWITCH_LABEL = "Enable kill switch"


@dataclass(frozen=True)
class KillSwitchConfirmDecision:
    """Result of evaluating whether a kill-switch opt-in change may persist."""

    allow_persist: bool
    next_opt_in: bool
    reason: str = ""


def evaluate_kill_switch_confirm(
    *,
    desired_on: bool,
    confirm_text: Optional[str] = None,
    cancelled: bool = False,
) -> KillSwitchConfirmDecision:
    """Pure gate for kill-switch Settings changes.

    - Turning **OFF** (``desired_on=False``): always allowed, no token needed.
    - Turning **ON** (``desired_on=True``): allowed only when ``confirm_text``
      equals :data:`KILL_SWITCH_CONFIRM_TOKEN` exactly (after outer trim).
    - Empty / wrong / cancelled ON attempts: ``allow_persist`` false, next off.
    """
    if not desired_on:
        return KillSwitchConfirmDecision(
            allow_persist=True,
            next_opt_in=False,
            reason="disable_no_confirm",
        )
    if cancelled:
        return KillSwitchConfirmDecision(
            allow_persist=False,
            next_opt_in=False,
            reason="enable_cancelled",
        )
    typed = (confirm_text or "").strip()
    if typed == KILL_SWITCH_CONFIRM_TOKEN:
        return KillSwitchConfirmDecision(
            allow_persist=True,
            next_opt_in=True,
            reason="enable_token_ok",
        )
    return KillSwitchConfirmDecision(
        allow_persist=False,
        next_opt_in=False,
        reason="enable_empty_token" if not typed else "enable_wrong_token",
    )


def kill_switch_confirm_token_matches(text: Optional[str]) -> bool:
    """True when *text* is the exact enable token (trim outer whitespace)."""
    return (text or "").strip() == KILL_SWITCH_CONFIRM_TOKEN
